New SMT2 backend: arithmetic operators implementation for fixed size bit vectors

NlightNFotis opened the pull request:
This adds implementations (and tests) for the binary arithmetic operators (and their conversion) to SMT2 data structures.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.


<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add implementation of arithmetic operator `bvadd`.

Also add tests for the data structure itself and the conversion of
`plus_exprt` to `smt_bit_vector_theoryt::addition`.

NlightNFotis committed:
Add implementation of multiplication operator `bvmul`.

Also add tests for its conversion from `mult_exprt` and of the factory
integrity.

NlightNFotis committed:
Add implementation of (unsigned) division operator `bvudiv`.

Along with tests of its conversion and the function application factory integrity.

NlightNFotis committed:
Add (unsigned) remainder (modulus) arithmetic operator `bvurem`.

Along with tests of its conversion and the function application factory integrity.

NlightNFotis committed:
Add implementation of negation (unary minus - `bvneg`) operator.

Along with conversion tests and tests of the factory method integrity.

TGWDB reviewed (approved):
Looks good. It would be nice to have unit tests of the validation of bitvector width being the same for the binary operations to ensure this is properly caught.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -686,12 +686,6 @@ smt_termt convert_expr_to_smt(const exprt &expr)
   {
     return convert_expr_to_smt(*float_not_equal);
   }
-  if(
```
🛠️ Could you add an explanation of why this would have incorrectly called `binary_relation_exprt` previously to the commit message body? As in - "The casting operations for `binary_relation_exprt` incorrectly allow casting to this type for expressions which do not derive from it".

thomasspriggs replied:
This is now resolved due to rebasing on develop.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
-    "Generation of SMT formula for plus expression: " + plus.pretty());
+  for(const auto &op : plus.operands())
+  {
+    PRECONDITION(can_cast_type<bitvector_typet>(op.type()));
```
⛏️ The way this is implemented leaves no place where the control flow "falls through" for non-bitvector types. This means that we have to refactor this implementation when we add extra support later. For example, the old `smt2_convt::convert_plus` also supports conversion of `ID_rational`, `ID_integer`, `ID_pointer` and `ID_vector` in addition to bitvectors. So we can expect to be adding these later.

NlightNFotis replied:
My rationale for this is that for now we do not support them (or more precisely, the operation is not supported for any type other than bitvectors) so I'm guarding for that type in the precondition. 

That check would be relaxed when needed (when support for the conversion of integers, rationals, etc are added).

Do you mean that you want a special message to that effect?

thomasspriggs replied:
I agree that this should have appropriate `INVARIANT`s. My argument is that this should be done based on an `UNIMPLMENTED` condition at the end of this function, rather than a `PRECONDITION` at the start. So the parts we have implemented are reached when conditions implemented are satisfied and we fall through to the `UNIMPLEMENTED` macro, if the given expression is not supported by any of the cases we have implemented.

My comment relates to how easy it will be to extend the implementation you have added, not really the ease of understanding it. Therefore I don't think additional messages would be helpful here.

thomasspriggs replied:
Looks good to me.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
+  const smt_termt &rhs)
+{
+  // For now, make sure that they have the same bit-width
+  additiont::validate(lhs, rhs);
```
🚫 The single caller of `return_sort` already calls `validate`. Adding this here means that it gets called twice. Remove this line please.

thomasspriggs replied:
Looks ok to me.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
+  // For now, make sure that they have the same bit-width
+  additiont::validate(lhs, rhs);
+  const auto width = lhs.get_sort().cast<smt_bit_vector_sortt>()->bit_width();
+  return smt_bit_vector_sortt{width};
```
🚫  Extracting the `bit_width` from the existing `bit_vectory_sortt`, just to construct a new sort which is identical to the original sort is doing unnecessary work and is preventing `irept` sharing from doing its thing. Just use `return lhs.get_sort();`

thomasspriggs replied:
This is now resolved.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,29 @@ static void validate_bit_vector_predicate_arguments(
+  const smt_termt &lhs,
+  const smt_termt &rhs)
+{
+  validate_bit_vector_predicate_arguments(lhs, rhs);
```
🚫 Calling a function named `validate_bit_vector_predicate_arguments` for validation of something which is not a predicate is highly misleading. I have no problem with you reusing the function, but it needs renaming in order to ensure the code remains readable.

thomasspriggs replied:
This still needs to be addressed.

thomasspriggs replied:
This comment has now been resolved by the rename commit.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,14 @@ class smt_bit_vector_theoryt
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<additiont> addition;
```
🚫 Rename `addition` to `add` please.

thomasspriggs replied:
Could you do the equivalent rename for the class as well please? `additiont` -> `addt`.

thomasspriggs replied:
This is now resolved.

thomasspriggs commented on unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -239,3 +239,58 @@ TEST_CASE(
+  }
+
+  SECTION(
+    "Addition of four counstant size bit-vectors - testing multi-ary handling "
```
⛏️ `counstant` is not an (English) word.

thomasspriggs replied:
Looks to be resolved.

thomasspriggs commented on unit/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -239,3 +239,58 @@ TEST_CASE(
+  {
+    const cbmc_invariants_should_throwt invariants_throw;
+    REQUIRE_THROWS(
+      convert_expr_to_smt(plus_exprt{true_exprt{}, false_exprt{}}));
```
💡 It would also be good to test that malformed instances of `plus_exprt` with 0 or 1 operands fail an invariant.

thomasspriggs replied:
Could we get the test for 1 operand as well as 0 operands?

thomasspriggs commented on unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -166,3 +166,24 @@ TEST_CASE("SMT bit vector predicates", "[core][smt2_incremental]")
+    REQUIRE(function_application.arguments().size() == 2);
+    REQUIRE(function_application.arguments()[0].get() == two);
+    REQUIRE(function_application.arguments()[1].get() == three);
+  }
```
⛏️ We should test that calling `smt_bit_vector_theoryt::add` with bit_vectors of mismatched width or on bool sorted terms fails an invariant.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for plus expression: " + plus.pretty());
+  for(const auto &op : plus.operands())
```
⛏️ `op` should be `operand` given the coding standards guideline to "avoid abbreviations". `op` is a particularly ambiguous abbreviation given that you need to read the context to know if it is an "operator" / "operation" or "operand".

thomasspriggs replied:
Looks good to me.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -686,12 +686,6 @@ smt_termt convert_expr_to_smt(const exprt &expr)
   {
     return convert_expr_to_smt(*float_not_equal);
   }
-  if(
```
⛏️ Can you keep commit titles short enough that they display in their entirety in the github UI without having to expand them?

thomasspriggs replied:
This looks to be resolved.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -325,8 +325,12 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 
 static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 {
-  UNIMPLEMENTED_FEATURE(
```
⛏️ I would prefer it if commits could be titled such that the title encompasses the entirety of what they change. The fact that you effectively have "Also add conversion of `plus_exprt` to `smt_bit_vector_theoryt::addition`" in the message body is kind of a red flag which suggests that you have combined unrelated changes together (even if one change depends of the other).

thomasspriggs reviewed (changes requested):
This an initial review of the first two commits titled - "Change order of casting and dispatching of binary_relation_exprt in `convert_expr_to_smt`" and "Add implementation of arithmetic operator bvadd." only, so that I can leave some initial comments. My review of the remaining commits is still to follow.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -335,8 +335,10 @@ static smt_termt convert_expr_to_smt(const plus_exprt &plus)
 
 static smt_termt convert_expr_to_smt(const minus_exprt &minus)
 {
-  UNIMPLEMENTED_FEATURE(
```
⛏️ This should still hit an UNIMPLEMENTED_FEATURE macro, when this function encounters types which are not yet implemented. This is because the implementation cannot be considered to be complete as it currently stands.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -68,3 +68,29 @@ void smt_bit_vector_theoryt::additiont::validate(
+  const smt_termt &rhs)
+{
+  // For now, make sure that they have the same bit-width
+  subtractiont::validate(lhs, rhs);
```
🚫 See previous comment on `::validate` for the plus operation.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -28,6 +28,15 @@ class smt_bit_vector_theoryt
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<subtractiont>
+    subtraction;
```
🚫 Rename `subtraction` to `subtract`, please.

thomasspriggs commented on unit/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -186,4 +186,17 @@ TEST_CASE(
+    REQUIRE(function_application.get_sort() == smt_bit_vector_sortt{8});
+    REQUIRE(function_application.arguments().size() == 2);
+    REQUIRE(function_application.arguments()[0].get() == two);
+    REQUIRE(function_application.arguments()[1].get() == three);
```
⛏️ Tests for invariant violation on constructing with an `lhs` and `rhs` of mismatched widths and terms of bool sort please.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<multiplicationt>
+    multiplication;
```
🚫 Rename `multiplication` to `multiply`, please.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
+    static smt_sortt return_sort(const smt_termt &lhs, const smt_termt &rhs);
+    static void validate(const smt_termt &lhs, const smt_termt &rhs);
+  };
+  static const smt_function_application_termt::factoryt<divisiont> division;
```
🚫 Rename `division` to `divide`, please.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
+    static smt_sortt return_sort(const smt_termt &operand);
+    static void validate(const smt_termt &operand);
+  };
+  static const smt_function_application_termt::factoryt<negationt> negation;
```
🚫 Rename `negation` to `negate`, please.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
+
+const char *smt_bit_vector_theoryt::subtractiont::identifier()
+{
+  return "bvsub";
```
❗ `bvsub` does not appear to be a valid operator in bit vector theory, according to this definition - https://smtlib.cs.uiowa.edu/theories-FixedSizeBitVectors.shtml . We should probably exclude it from our definition of the theory and convert minus expressions to a combination of plus and negation terms instead.

thomasspriggs replied:
Please disregard this comment, as `bvsub` is defined as part of `logic QF_BV`, rather than as part of `theory FixedSizeBitVectors`

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
+
+const char *smt_bit_vector_theoryt::divisiont::identifier()
+{
+  return "bvdiv";
```
❗ This should be `bvudiv`.

thomasspriggs replied:
or `bvsdiv` as appropriate.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.cpp:
```diff
@@ -42,3 +42,152 @@ static void validate_bit_vector_predicate_arguments(
+}
+
+void smt_bit_vector_theoryt::negationt::validate(const smt_termt &operand)
+{
```
🚫 Missing validation. This should check that the operand has bitvector sort.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -154,9 +154,9 @@ static smt_termt convert_expr_to_smt(const bitnot_exprt &bitwise_not)
 
 static smt_termt convert_expr_to_smt(const unary_minus_exprt &unary_minus)
 {
-  UNIMPLEMENTED_FEATURE(
```
:no_entry_sign: This `UNIMPLEMENTED_FEATURE` should remain and be reachable, because conversion for types including `ID_rational`, `ID_integer`, `ID_real` and `ID_floatbv` are unimplemented. This comment also applies to lines of code with a :bird: comment.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -154,9 +154,9 @@ static smt_termt convert_expr_to_smt(const bitnot_exprt &bitwise_not)
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for unary minus expression: " +
-    unary_minus.pretty());
+  PRECONDITION(can_cast_type<bitvector_typet>(unary_minus.op().type()));
```
:no_entry_sign: General comment on the usage of `can_cast_type<bitvector_typet>` / `type_try_dynamic_cast<bitvector_typet>` - There are sub types of `bitvector_typet` for which the conversions which you have defined are not the correct behaviour and were not intended to be included in the current piece of work. These include `fixedbv_typet`, `floatbv_typet`, and `pointer_typet`. If expressions of these types are passed into the new expression to smt conversion then we must fail an `UNIMPLEMENTED` invariant rather than emitting the incorrect SMT conversion.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -325,20 +325,28 @@ convert_expr_to_smt(const binary_relation_exprt &binary_relation)
 static smt_termt convert_expr_to_smt(const div_exprt &divide)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for divide expression: " + divide.pretty());
```
:bird: See previous comment on unimplemented features.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -367,8 +377,12 @@ convert_expr_to_smt(const euclidean_mod_exprt &euclidean_modulo)
 static smt_termt convert_expr_to_smt(const mult_exprt &multiply)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for multiply expression: " + multiply.pretty());
```
:bird: See previous comment on unimplemented features.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -352,9 +360,11 @@ static smt_termt convert_expr_to_smt(const ieee_float_op_exprt &float_operation)
 {
-  UNIMPLEMENTED_FEATURE(
-    "Generation of SMT formula for truncation modulo expression: " +
-    truncation_modulo.pretty());
```
:pick: We expect this operator to be applied only to (signed/unsigned) integer bit vectors. It might be good to document this expectation using a `UNEXPECTEDCASE` macro, similar to how `smt2_convt::convert_mod` does this.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
+  };
+  static const smt_function_application_termt::factoryt<divisiont> division;
+
+  struct remaindert final
```
:no_entry_sign: We actually need 2 variations of this operation because there is the unsigned remainder `bvurem` and the signed remainder `bvsrem`.

thomasspriggs commented on src/solvers/smt2_incremental/smt_bit_vector_theory.h:
```diff
@@ -20,6 +20,56 @@ class smt_bit_vector_theoryt
+  static const smt_function_application_termt::factoryt<multiplicationt>
+    multiplication;
+
+  struct divisiont final
```
:no_entry_sign: We actually need 2 variations of this operation because there is the unsigned divide `bvudiv` and the signed divide `bvsdiv`.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -326,8 +326,21 @@ static optionalt<smt_termt> try_relational_conversion(const exprt &expr)
+       [](exprt operand)
+       { return can_cast_type<bitvector_typet>(operand.type()); }))
+  {
+    // if all the operands are bitvectors, call the bitvector addition constructor
```
:pick: I don't consider this comment to be useful. This is because it states what can already been seen from reading the code, not why or how it does it.

NlightNFotis committed:
Add implementation of arithmetic operator `bvsub`.

Also add tests for it, both the conversion of `minus_exprt` to it,
and the node factory constructing the nodes appropriately.

NlightNFotis committed:
Rename `validate_bit_vector_predicate_arguments` to `validate_bit_vector_operator_arguments`.

This allows the name to reflect the more general purpose deployment of this function.

NlightNFotis committed:
Add implementation of (signed) division operator (`bvsdiv`).

In addition, modify the tests and the implementation of the conversion function
to account for signed/unsigned input.

NlightNFotis committed:
Add implementation of signed remainder operator (`bvsrem`).

Along with this, change the tests and implementation of the conversion
function to account for signed/unsigned inputs.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -366,10 +366,22 @@ static smt_termt convert_expr_to_smt(const div_exprt &divide)
     can_cast_type<integer_bitvector_typet>(divide.lhs().type()) &&
     can_cast_type<integer_bitvector_typet>(divide.rhs().type());
 
+  bool both_operands_unsigned =
```
:pick: `const` - Every variable which is not mutated should be declared `const`. I have picked this as an arbitrary example, but I think it is missing in a few places.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -366,10 +366,22 @@ static smt_termt convert_expr_to_smt(const div_exprt &divide)
+    else
+    {
+      return smt_bit_vector_theoryt::signed_divide(
+        convert_expr_to_smt(divide.lhs()), convert_expr_to_smt(divide.rhs()));
```
:bulb: I suggest introducing `lhs_term` and `rhs_term` variables above this if statement, so that the calls to `convert_expr_to_smt` are not duplicated in the two branches.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -401,9 +401,24 @@ static smt_termt convert_expr_to_smt(const ieee_float_op_exprt &float_operation)
+  }
+  else
+  {
+    {
```
:pick: Superfluous brackets.

thomasspriggs commented on src/solvers/smt2_incremental/convert_expr_to_smt.cpp:
```diff
@@ -405,11 +405,24 @@ static smt_termt convert_expr_to_smt(const mod_exprt &truncation_modulo)
+    {
+      return smt_bit_vector_theoryt::signed_remainder(
+        convert_expr_to_smt(truncation_modulo.lhs()),
+        convert_expr_to_smt(truncation_modulo.rhs()));
```
:bulb: Factor out the common `convert_expr_to_smt` sub-expressions, like I suggested with your implementation for the divide operations.

thomasspriggs reviewed (approved):
I am now approving as all of my :no_entry_sign: comments have been addressed.
